### PR TITLE
Feature/particle tuple

### PR DESCRIPTION
**Pull request by whaeck**
Removed the JoinedIdentifier since it did not provide a lot of flexibility. Instead, I added a free function to join and split strings (tests were added as well). ParticleTupleID was modified accordingly.

Tests were added for bad data for every identifier type

### CODE REVIEW

File: src/elementary/src/join.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    if ( not strings.size() ) {
+
+      return "";
+    }
+
+    auto iter = strings.begin();
+    auto end = strings.end();
+    std::string start = *iter++;
+    return std::accumulate( iter, end, start,
```
Comment (jlconlin) [inline src/elementary/src/join.hpp:30]: :+1

File: src/elementary/src/join.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+namespace njoy {
+namespace elementary {
+
+  /**
+   *  @brief Join a list of string to a single string using a given separator
+   *
+   *  @param[in] strings     the strings to be joined together
+   *  @param[in] separator   the separator to be used
+   */
+  std::string join( const std::vector< std::string >& strings,
```
Comment (jlconlin) [inline src/elementary/src/join.hpp:19]: I like this `join` function. We should (at some point) move this into [utility](https://github.com/njoy/utility).

File: src/elementary/src/split.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+namespace njoy {
+namespace elementary {
+
+  /**
+   *  @brief Split a string into a list of strings using a given separator
+   *
+   *  @param[in] string      the strings to be split
+   *  @param[in] separator   the separator to be used
+   */
+  std::vector< std::string > split( const std::string& string,
```
Comment (jlconlin) [inline src/elementary/src/split.hpp:19]: We have a `split` function in [utility](https://github.com/njoy/utility), but I acknowledge that you don’t want to have any dependencies with this.

File: src/elementary/test/elementary.test.cpp [hunk 4]
```diff
@@ -59,3 +61,70 @@ SCENARIO( "emit" ) {
...
   } // GIVEN
 } // SCENARIO
+
+SCENARIO( "join" ) {
+
+  GIVEN( "a vector of strings" ) {
+
+    std::vector< std::string > empty = {};
+    std::vector< std::string > one = { "ab" };
+    std::vector< std::string > strings = { "ab", "cd", "ef" };
```
Comment (jlconlin) [inline src/elementary/test/elementary.test.cpp:71]: It makes me sad you couldn’t come up with something more fun than this. There are so many things you could have done instead, like favorite muppet characters, people who annoy you, worst European cities, etc.

Comment (whaeck) [reply src/elementary/test/elementary.test.cpp:71]: At least I know my ABCs ;-)

**Pull request merged**